src/pykeen/evaluation/rank_based_evaluator.py

```diff
@@ -320,3 +321,52 @@ def finalize(self) -> RankBasedMetricResults:  # noqa: D102
             hits_at_k=dict(hits_at_k),
             adjusted_mean_rank=adjusted_mean_rank,
         )
+
+
+def prepare_triple_rank_df(
+    mapped_triples: MappedTriples,
+    ranks: Mapping[Tuple[str, str], Sequence[float]],
+) -> Tuple[pandas.DataFrame, pandas.DataFrame]:
+    """
+    Prepare a pair of dataframes for extended evaluation of ranks.
+
+    :param mapped_triples: shape: (num_triples, 3)
+        The ID-based triples.
+    :param ranks:
+        The ranks, a mapping from (side, rank_type) to a sequence of ranks. Each sequence has the same length:
+        `num_triples`.
+
+    .. warning ::
+        Important: The mapped triples and the values of the ranks have to be in a consistent order.
+
+
+    Usage example: Compute mean rank individually for each relation type:
+
+    >>> tdf, rdf = prepare_triple_rank_df(mapped_triples, ranks)
```

Comment:
could you extend this example to work on the results returned by
```python
from pykeen.pipeline import pipeline
results = pipeline(dataset='Nations', model='RotatE')
...
```

so it's a full example?

Reply:
Right now, no, because the individual ranks get discarded when the aggregated results are extracted.

Reply:
okay yeah that's what I thought because we're trying to avoid this memory leak. do you think there might be a way around this issue?

Reply:
I do not think that this is something which needs to be in the pipeline, since most of the time you will not use it. Since we have the model in the result, we could evaluate this model after finishing the pipeline, this time keeping the individual ranks. Maybe we can add a flag to the `RankBasedEvaluator`'s `__init__` to make discarding optional.

Reply:
Another option would be to modify the evaluators' API:
Right now it is stateful an relies on clearing all caches when finalize is called, which is done at least once at the end of an evaluation round.

We could  change that to create a new object (tentative name "aggregator") which only lives for the time of one evaluation round. We can keep a reference to the last one, if we want to (at the cost of *one* aggregation object's memory, instead of the unbound memory requirement of the previous one, if finalize did not clear caches).